src/transports/quic.rs

```diff
@@ -0,0 +1,180 @@
+use mio::{net::UdpSocket, Events, Interest, Poll, Token, Waker};
+use quiche::{accept, connect, ConnectionId};
+
+use crate::{
+    endpoint::Endpoint, error::PeerNetError, network_manager::SharedPeerDB, peer::Peer,
+    peer_id::PeerId,
+};
+
+use super::Transport;
+
+const NEW_CONNECTION_SERVER: Token = Token(0);
+const STOP_LISTENER: Token = Token(10);
+
+pub(crate) struct QuicTransport {
+    pub peer_db: SharedPeerDB,
+    pub out_connection_attempts: WaitGroup,
+    pub listeners: HashMap<SocketAddr, (Waker, JoinHandle<()>)>,
+}
+
+#[derive(Clone)]
+pub struct QuicOutConnectionConfig {
+    pub identity: PeerId,
+    pub local_addr: SocketAddr,
+}
+
+impl QuicTransport {
+    pub fn new(peer_db: SharedPeerDB) -> QuicTransport {
+        QuicTransport {
+            peer_db,
+            out_connection_attempts: WaitGroup::new(),
+            listeners: Default::default(),
+        }
+    }
+}
+
+impl Transport for QuicTransport {
+    type OutConnectionConfig = QuicOutConnectionConfig;
+
+    fn start_listener(&mut self, address: SocketAddr) -> Result<(), PeerNetError> {
+        let mut poll = Poll::new().map_err(|err| PeerNetError::ListenerError(err.to_string()))?;
+        let mut events = Events::with_capacity(128);
+        let waker = Waker::new(poll.registry(), STOP_LISTENER)
+            .map_err(|err| PeerNetError::ListenerError(err.to_string()))?;
+        let listener_handle = std::thread::spawn({
+            let peer_db = self.peer_db.clone();
+            move || {
+                let mut socket = UdpSocket::bind(address)
+                    .expect(&format!("Can't bind TCP transport to address {}", address));
+                // Start listening for incoming connections.
+                poll.registry()
+                    .register(&mut socket, NEW_CONNECTION_SERVER, Interest::READABLE)
+                    .expect(&format!(
+                        "Can't register polling on TCP transport of address {}",
+                        address
+                    ));
+                let mut buf = [0; PUBLIC_KEY_SIZE_BYTES + SIGNATURE_SIZE_BYTES];
+                loop {
+                    println!("Listening for new connections...");
+                    // Poll Mio for events, blocking until we get an event.
+                    poll.poll(&mut events, None)
+                        .expect(&format!("Can't poll QUIC transport of address {}", address));
+
+                    // Process each event.
+                    for event in events.iter() {
+                        match event.token() {
+                            NEW_CONNECTION_SERVER => {
+                                //TODO: Error handling
+                                let (_num_recv, from_addr) = socket.recv_from(&mut buf).unwrap();
+                                let mut config =
+                                    quiche::Config::new(quiche::PROTOCOL_VERSION).unwrap();
+                                accept(
```

Comment:
It's opinionated, I like having something like `quiche::accept`. It's easier to review

Reply:
Agree. Done